doc/designs/quic-design/congestion-control.md

```diff
@@ -0,0 +1,229 @@
+Congestion control API design
+=============================
+
+This is mostly inspired by the MSQUIC congestion control API
+as it was the most isolated one among the libraries.
+
+The API is designed in a way to be later easily transformed into a
+a fetchable implementation API.
+
+The API is centered around two structures - the `OSSL_CC_METHOD`
+structure holding the API calls into the congestion control module
+and `OSSL_CC_DATA` opaque type that holds the data of needed
+for the operation of the module.
+
+Most of the information when the functions of the API are
+supposed to be called and how the API is designed to operate
+should be clear from the comments for the individual
+functions in [Appendix A](#appendix-a).
+
+Changeables
+-----------
+
+As some parameters that the congestion control algorithm needs might
+be updated during the lifetime of the connection these parameters
+are called changeables.
+
+The CC implementation will save the pointers to these parameters'
+data and will use the current value of the data at the pointer when
+it needs to recalculate the allowance and whether the sending is
+blocked or not.
+
+Some examples of changeables
+
+ - `payload_length` - the maximum length of payload that can be sent
+   in a single UDP packet (not counting UDP and IP headers).
```

Comment:
Unless we support sending over multiple-paths simultaneously the value for the active path is what matters. I suppose once we support multiple simultaneous paths the OSSL_CC_DATA would be per-path instead of per-connection.